README.md
<div align="center">
    <h1>POP3 Client (29) </h1>
    <h3>fischer_project_2</h3>
    <h5>David Fischer, 5CHIF (03)</h5>
</div>

<div align="center">

[![License](https://img.shields.io/badge/License-Boost%201.0-blue.svg)](https://www.boost.org/LICENSE_1_0.txt)
[![time tracker](https://wakatime.com/badge/github/konstfish/fischer_project_2.svg)](https://wakatime.com/badge/github/konstfish/fischer_project_2)
[![CodeFactor](https://www.codefactor.io/repository/github/konstfish/fischer_project_2/badge?s=c397d4ac5a23901eaf68a5ace37595352a91e3d1)](https://www.codefactor.io/repository/github/konstfish/fischer_project_2)

</div>

## 💦 Short Summary
The goal of this assignment was the creation of a POP3 Client, including support for GnuTLS. The project was implemented using Protobuf, as well as gRPC.

```
  _           _     _     _        0  
  | |__  _   _| |__ | |__ | | ___    o   
  | '_ \| | | | '_ \| '_ \| |/ _ \    O 
  | |_) | |_| | |_) | |_) | |  __/   0  
  |_.__/ \__,_|_.__/|_.__/|_|\___|    o         
```

## 📈 Usage

```
POP3 Client
Usage: ./bubble [OPTIONS]

Options:
  -h,--help                   Print this help message and exit
  -s,--server TEXT            Server Domain
  -u,--user TEXT              Account Name
  -p,--pass TEXT              Account Password
  -t,--tls                    Connect using TLS
  --port INT                  Specify custom port
  -j,--json TEXT              Relative path to JSON file containing the Config
  -i,--interactive            Enable Interactive Shell
```

## 📝 Requirements

* [Protobuf](https://developers.google.com/protocol-buffers)
* [gRPC](https://grpc.io/)

- - -

* [CLI11](https://github.com/CLIUtils/CLI11)
* [JSON for Modern C++](https://github.com/nlohmann/json)
* [spdlog](https://github.com/gabime/spdlog)
* [tabulate](https://github.com/p-ranav/tabulate)
* [httplib](https://github.com/yhirose/cpp-httplib)
* [inja](https://github.com/pantor/inja)

## 🔨 Compile

`mkdir -p build && cd build && meson .. && ninja`

## 🖥️ Example Interactive Client
```
fischer_project_2/build on  master [!] at ☸️  docker-desktop took 23s
➜ ./bubble -j ../config.json -i
[2021-03-14 20:24:36.812] [console] [info] Connection successfully established
[2021-03-14 20:24:37.723] [console] [info] Successfully Logged in!
[2021-03-14 20:24:37.725] [console] [info] gRPC Server listening on 0.0.0.0:50051
bubble> help
Bubble Interactive Client:
dl / download <message_id>  - download email with id
rm / delete <message_id>    - delete email with id
ls / list <amount>          - list amount of mails
exit                        - disconnects the session
help                        - displays this message
bubble> ls 5
+------------+-----------------------------------+------------------------------------------------------------------+---------------------------------+
| Message ID | Recieved From                     | Subject                                                          | Date                            |
+------------+-----------------------------------+------------------------------------------------------------------+---------------------------------+
| 6          | David Fischer <david@konst.fish>  | Fwd: Philip Trauner und andere teilen ihre Meinung auf LinkedIn  | Sun, 7 Mar 2021 21:13:56 +0100  |
+------------+-----------------------------------+------------------------------------------------------------------+---------------------------------+
| 5          | David Fischer <david@konst.fish>  | Fwd: Sie werden wahrgenommen, David Fischer                      | Sun, 7 Mar 2021 21:13:53 +0100  |
+------------+-----------------------------------+------------------------------------------------------------------+---------------------------------+
| 4          | David Fischer <david@konst.fish>  | Fwd: DJI MSDK iOS Update Delay Notice                            | Sun, 7 Mar 2021 21:13:46 +0100  |
+------------+-----------------------------------+------------------------------------------------------------------+---------------------------------+
| 3          | David Fischer <david@konst.fish>  | Fwd: Hot in March                                                | Sun, 7 Mar 2021 21:13:13 +0100  |
+------------+-----------------------------------+------------------------------------------------------------------+---------------------------------+
| 2          | David Fischer <david@konst.fish>  | Fwd: WIRED und andere teilen ihre Meinung auf LinkedIn           | Sat, 6 Mar 2021 22:59:52 +0100  |
+------------+-----------------------------------+------------------------------------------------------------------+---------------------------------+
bubble> dl 6
[2021-03-14 20:24:47.324] [console] [info] Saved message with ID: 6
bubble> rm 6
[2021-03-14 20:24:51.523] [console] [info] Deleted message with ID: 6
bubble> exit
[2021-03-14 20:24:54.536] [console] [info] Successfully Quit Session!
```

## 🌐 Web Interface

![web interface](/doc/images/web_interface.png?raw=true)



CHANGELOG.org
* Changelog
** [Unreleased]
*** Added
- Create repository
** [1.0.0] - 2020-04-05
*** Added
- Correct grammatical errors in Documentation 
- Update README.md
- Ready up for 1.0.0 Release
** [0.0.6] - 2020-03-29
*** Added
- POP3CSImplementation.h & POP3CSClient.h Files
- Add http_server folder
- Add http_server meson build executable
- Add main.cpp to http_server sources
- Add POP3CSClient to http_server src & include
- Include httplib (for http_server)
- Inlcude inja (for http_server)
- Add httplib server definition to http_server main.cpp
- Add templates folder ( & render.html)
- Add gRPC handling to http_server
- Add retrieve_message_meta_proto_dry function to ProtoInterface Class
- Add http_server & templates folder to Documentation
- Add httplib Section to Documentation
- Add inja Section to Documentation
- Add #pragma GCC diagnostic includes to gRPC
- Add comments to every function
- Add comments to classes
- Delete unused code & Debug
- Add Classes to Documentation
*** Changed
- Split ProtoInterface.h into seperate Files
- Update render.html File
- Add dry run to ProtoInterface::retrieve_messages function
- Change json include
- Update Message Lister to function with longer metadata
- Update main.cpp & Interactive.h to enable compilation on Linux Hosts
*** Removed
- Deprecate access to POP3 Functions using CL Arguments
** [0.0.5] - 2020-03-15
*** Added
- Add protobuff test directorypop3.proto to src
- Add protobuff build to meson
- Add ProtoInterface class
- Add ProtoBuff handler functions
- Include grpc Library
- Add grpc Server Class
- Add grpc Client Class
- Add grpc routes to pop3.proto
- Add grpc routes to Server Class
- Add grpc routes to Client Class
- Add error handling to JSON Parser
- Add Introduction Section to Documentation
- Add References to Documentation
- Add Class Diagram to Documentation
- Add Implementation Section to Documentation
- Add Code Section to Documentation
- Add Library Section to Documentation
- Add grpc/protobuf Sections to Documentation
- Add Usage Section to Documentation
- Add Project Structure Section to Documentation
- Add Usage example to README.md
*** Changed
- Update meson.build file
- Add a few sections to README.md
- Update meson.build to feature grpc
- Rewrote Interactive Class to use gRPC client as communication medium
- Clean up ProtoInterface.h
- Move Console spdlog entries to Interactive class
- Implement changes from "Zwischenstand"
** [0.0.4] - 2020-03-08
*** Added
- Add basic socket to pop3client Class
- Add basic resolver to pop3client Class
- Add read and write methods to pop3client Class
- Add GnuTLS support to pop3client Class (Setup, read/write methods)
- Add utils files for support classes
- Add write_to_file method to utils class, to enable mails to be written as .eml files
- Add documentation TeX structure & references File
- Include spdlog Library
- Add basic spdlog file sink
- Add basic error handling to establish_connection function
- Add split functions to utils.cpp
- Add delete_message function to pop3client class
- Add get_total_messages function to pop3client class
- Add retrieve_messages function to pop3client class
- Add login function to pop3client class
- Include tablulate Library in util.h
- Add quit funtion to pop3client class (gracefully ends sessions)
- Add ends_with funtion to utils 
- Add read_to_end function to pop3client class (reads messages larger than 8000 chars)
- Include CLI11 Library in main.cpp
- Add simple CLA handling
- Include JSON Library in main.cpp
- Add simple json parsing to read config file
- Add simple commands to CLA list
- Add simple handling for CLA commands
- Add Interactive.h & Interactive.cpp
- Add Interactive class to handle responsive input of commands
- Add CLA to enable Interactive shell
*** Changed
- Split temp function into establish_connection & debug functions
- Fixed output in pop3client_utils::print_messages
- Cleaned up save_mail funtion (now saves complete emails as .eml files with the emails subject as file name)
- Update Naming Scheme
- Update retrieve_messages function. Now works when messages are marked for deletion
- Update Naming Scheme
- Update retrieve_messages function. Now works when messages are marked for deletion
*** Removed
- Remove subprocess attempt
- Remove backup folder, as it no longer holds any importance
** [0.0.3] - 2020-03-01
*** Added
- Add Time Tracking Badge to README
*** Changed
- Moved asio approach to backup folder
- Made multiple attempts to work with subprocesses, to no proceeds
** [0.0.2] - 2020-02-22
*** Changed
- Rewrote Socket to feature a resolver
- Made basic changes to prepare TLS sockets
** [0.0.1] - 2020-02-15
*** Added
- Include ASIO Library in main.cpp
- Fill .gitignore
- Add simple interfacing with local POP3 server to main.cpp
- Create RESEARCH.md to temporarily document sources
- Create empty doc file


src/main.cpp
/*
author: David Fischer
matnr:  i16079
file:   main.cpp
desc:   NVS Projekt 2 - Simpler POP3 Client mit TLS Support
class:  5C
catnr:  03
*/

#include "POP3client.h"
#include "Interactive.h"
#include "ProtoInterface.h"
#include "POP3CSImplementation.h"

#include "CLI11.hpp"
#include <json.hpp>

#include <spdlog/spdlog.h>
#include <spdlog/sinks/stdout_color_sinks.h>
#include "spdlog/sinks/basic_file_sink.h"

#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wunused-parameter"
#pragma GCC diagnostic ignored "-Wnon-virtual-dtor"
#include <grpc/grpc.h>
#include <grpcpp/server.h>
#include <grpcpp/server_builder.h>
#include <grpcpp/server_context.h>
#pragma GCC diagnostic pop

#include <iostream>

using namespace std;
using json = nlohmann::json;

int main(int argc, char* argv[]) {
    // CLI11 setup
    CLI::App app("POP3 Client");

    // Connection Options
    string servername{""};
    app.add_option("-s,--server", servername, "Server Domain");

    string username{""};
    app.add_option("-u,--user", username, "Account Name");

    string password{""};
    app.add_option("-p,--pass", password, "Account Password");

    bool use_tls{false};
    app.add_flag("-t,--tls", use_tls, "Connect using TLS");

    int port{-1};
    app.add_option("--port", port, "Specify custom port");

    string jsonfile{""};
    app.add_option("-j,--json", jsonfile, "Relative path to JSON file containing the Config");

    // Deprecated
    /*
    int download{-1};
    app.add_option("-d,--download", download, "Download Email with specified ID");

    int del{-1};
    app.add_option("-r,--remove", del, "Delete Email with specified ID");

    int list{-1};
    app.add_option("-l,--list", list, "List ")
        ->check(CLI::Range(1,50).description("Range of Emails limited to sensible values").active(true).name("range"));
    */

    // Enable interactive Shell
    bool enable_interactive{false};
    app.add_flag("-i,--interactive", enable_interactive, "Enable Interactive Shell");    

    // set up spdlog
    auto console = spdlog::stdout_color_mt("console");
    auto logger = spdlog::basic_logger_mt("logger", "logs/basic-log.txt");

    // PARSE argv
    try {
        app.parse(argc, argv);
    } catch (const CLI::ParseError &e) { 
        return app.exit(e);
    }

    // Prase json, if specified
    if(jsonfile != "") {
        try {
            std::ifstream i(jsonfile);
            json j;
            i >> j;
            //cout << j << endl;

            servername = j["hostname"];
            username = j["username"];
            password = j["password"];
            use_tls = j["tls"];
            port = j["port"];
        }catch(exception& e){
            spdlog::get("console")->error("Unable to read specified JSON File");
            return 1;
        }
    }

    // set port dynamically if not specified
    if(port == -1){
        if(use_tls){
            port = 995;
        }else{
            port = 110;
        }
    }

    spdlog::get("logger")->info("Starting POP3 Client");

    //pop3client c("fortimail.konst.fish", 110, false);
    //pop3client c("fortimail.konst.fish", 995, true);

    // create pop3client
    POP3client c(servername, port, use_tls);
    int check = 0;
    
    check = c.establish_connection();
    if(check){
        return 1;
    }

    check = c.login(username, password);
    if(check){
        return 1;
    }

    // check for commands
    // deprecated
    /*
    if(download != -1){
        c.save_mail(download);
    }

    if(list != -1){
        c.retrieve_messages(list);
    }

    if(del != -1){
        c.delete_message(del);
    }
    */

    // ProtoServer

    ProtoInterface proto(ref(c));

    // Initialize gRPC Server in Lambda Thread
    std::thread grpc_serv([&](){
        POP3CSImplementation service(ref(proto));
        std::string server_address("0.0.0.0:50051");

        grpc::ServerBuilder builder;
        builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
        builder.RegisterService(&service);
        std::unique_ptr<grpc::Server> server(builder.BuildAndStart());
        spdlog::get("console")->info("gRPC Server listening on {}", server_address);
        server->Wait();
    });

    // INTERACTIVE SHELL
    usleep(10000);
    
    if(enable_interactive){
        std::string address("localhost:50051");
        POP3CSClient client(
            grpc::CreateChannel(
                address, 
                grpc::InsecureChannelCredentials()
            )
        );

        Interactive shell(ref(client));
        shell.run();
    }

    // join grpc server
    grpc_serv.join();

    // end pop3client session
    c.quit();

    return 0;
}